src/rest_vol_dataset.c

```diff
@@ -4137,4 +4164,157 @@ rv_dataset_write_cb(hid_t mem_type_id, hid_t mem_space_id, void *buf, struct res
 {
     herr_t ret_value = SUCCEED;
     return SUCCEED;
-}
\ No newline at end of file
+}
+
+/*-------------------------------------------------------------------------
+ * Function:    RV_dataspace_selection_is_contiguous
+ *
+ * Purpose:     Checks if the specified dataspace in a contiguous selection.
+ *
+ * Return:      TRUE or FALSE if the selection is contiguous or
+ *              non-contiguous and FAIL if it is unable to determine it.
+ *
+ * Programmer:  Jan-Willem Blokland
+ *              August, 2023
+ */
+static htri_t
+RV_dataspace_selection_is_contiguous(hid_t space_id)
+{
+    htri_t   ret_value = TRUE;
+    hbool_t  whole     = TRUE;
+    hsize_t *dims      = NULL;
+    hsize_t *start     = NULL;
+    hsize_t *stride    = NULL;
+    hsize_t *count     = NULL;
+    hsize_t *block     = NULL;
+    int      i;
+    int      ndims;
+    hssize_t npoints, nblocks;
+
+    if ((npoints = H5Sget_select_npoints(space_id)) < 0)
+        FUNC_GOTO_ERROR(H5E_DATASPACE, H5E_CANTGET, FAIL, "can't retrieve number of selected points");
+    if (npoints < 2)
+        FUNC_GOTO_DONE(TRUE);
+
+    if ((ndims = H5Sget_simple_extent_ndims(space_id)) < 0)
+        FUNC_GOTO_ERROR(H5E_DATASPACE, H5E_CANTGET, FAIL, "can't retrieve dataspace dimensionality");
+    if (!ndims)
+        FUNC_GOTO_DONE(TRUE);
+
+    if (H5S_SEL_HYPERSLABS == H5Sget_select_type(space_id)) {
```

Comment:
This function will report that any point selection is contiguous.

Reply:
Yes, this is right. That was the original implementation.

Reading at the documentation of `H5Sselect_elements()` it is not complete clearly to me how it works. Looking at the given 1D example, should I read it like one selects 1st, 14th, 17th, 23rd and 8th element of 1D array which holds the data at these points. When calling `H5Dwrite()` this data array should have a length of at least 23 elements. If this is the correct way of looking at it then any point selection should be non-contiguous. In this case the correct data will be written to the right location. Of course it is sub-optimal for the case when the points are continuous in memory. I can implement this like this or is it desired to also check if the selected points are contiguous in memory to possible overhead caused by additional the memory copies.

Reply:
I believe you're right. When using a point selection, the memory buffer and memory dataspace should be contiguous (see Example 7 [here](https://davis.lbl.gov/Manuals/HDF5-1.8.7/UG/12_Dataspaces.html)) even if the file selection isn't, so this behavior makes sense.

Reply:
I propose we do the following to resolve the point selection case:
1. Implement any point selection is non-contiguous such this PR can be merged.
2. Next, I create a separate PR in which I improve the point selection case by taking into account if it is continuous and if so compute the corresponding offset in memory. Similar as the hyperslab case.

Reply:
Good idea. I spoke with Jordan about this, and most point selections are not contiguous in memory, so it's safest to assume they are non-contiguous and perform the memory management in `H5Dgather`

One thing about your previous example - when calling `H5Dwrite()`, the data array would only need a length of 5 elements, to hold the 1st, 14th, 17th, 23rd, and 8th elements.

Reply:
Thanks for the information. It makes me realize that I have to take a closer at the point selection to get a better understanding how it actually works.

In my latest commit, I have implemented the assumption that any point selection is non-contiguous in memory.